### PR TITLE
Implement proper stability check for `const impl Trait`, fall back to unstable `const` when undeclared

### PR DESCRIPTION
**Pull request by jhpratt**
<sup>Doing these as one PR because it fit with the way I was writing code. I can split them into separate commits if desired, though there's not much point.</sup>

Rather than deferring to const eval for checking if a trait is const, we now check up-front. This allows the error to be emitted earlier, notably at the same time as other stability checks.

Also included in this PR is a change of the default const stability level to **UN**stable. Previously, an item that was `const` but did not explicitly state it was unstable was implicitly stable.

r? @oli-obk 

@rustbot label +A-const-fn +A-stability +C-enhancement +F-const-trait-impl +S-waiting-on-review +T-compiler -T-rustdoc

### CODE REVIEW

File: compiler/rustc_attr/src/builtin.rs [hunk 1]
```diff
@@ -98,6 +98,15 @@ pub struct Stability {
     pub feature: Symbol,
 }
 
+impl Stability {
```
Comment (jhpratt) [inline compiler/rustc_attr/src/builtin.rs:101]: I've added some helper methods to help with potential ambiguity for readers. This way it's immediately clear if it's regular stability or const stability being checked.

File: compiler/rustc_const_eval/src/transform/check_consts/mod.rs [hunk 2]
```diff
@@ -98,22 +96,39 @@ pub fn is_const_stable_const_fn(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
...
-        None => {}
+        Some(stab) => stab.is_const_stable(),
+        None if is_parent_const_stable_trait(tcx, def_id) => {
+            // Remove this when `#![feature(const_trait_impl)]` is stabilized,
+            // returning `true` unconditionally.
+            tcx.sess.delay_span_bug(
+                tcx.def_span(def_id),
+                "trait implementations cannot be const stable yet",
+            );
+            true
```
Comment (jhpratt) [inline compiler/rustc_const_eval/src/transform/check_consts/mod.rs:110]: While this should never be countered, compilation will fail if it is. So I figured it's best to minimize the error messages here, as we already have a bug. Plus `true` is the eventual value we want.

File: compiler/rustc_interface/src/passes.rs [hunk 3]
```diff
@@ -979,6 +979,10 @@ fn analysis(tcx: TyCtxt<'_>, (): ()) -> Result<()> {
 
     sess.time("layout_testing", || layout_test::test_layout(tcx));
 
+    sess.time("stable_impl_const_trait_checking", || {
+        rustc_passes::stability::check_const_impl_trait(tcx)
+    });
```
Comment (jhpratt) [inline compiler/rustc_interface/src/passes.rs:984]: Where should this statement go? If it's any lower, then it won't run at the "correct" time.

File: compiler/rustc_passes/src/stability.rs [hunk 4]
```diff
@@ -655,6 +653,44 @@ impl<'tcx> Visitor<'tcx> for MissingStabilityAnnotations<'tcx> {
...
+            hir::ItemKind::Impl(hir::Impl {
+                of_trait: Some(_),
+                constness: hir::Constness::Const,
+                ..
+            })
+        ) {
+            return;
+        }
+
+        if self.tcx.lookup_const_stability(item.def_id).map_or(false, |stab| stab.is_const_stable())
```
Comment (jhpratt) [inline compiler/rustc_passes/src/stability.rs:673]: I wasn't sure — is using `tcx.lookup_const_stability()` or `tcx.has_attr()` preferred? In this situation I don't believe there's a semantic difference.

File: compiler/rustc_passes/src/stability.rs [hunk 5]
```diff
@@ -886,6 +922,17 @@ impl<'tcx> Visitor<'tcx> for CheckTraitImplStable<'tcx> {
     }
 }
 
+pub fn check_const_impl_trait(tcx: TyCtxt<'_>) {
+    let features = tcx.features(); // FIXME How cheap is this call?
```
Comment (jhpratt) [inline compiler/rustc_passes/src/stability.rs:926]: How cheap is `tcx.features()`? Should I just inline this variable?

File: src/test/ui/rfc-2632-const-trait-impl/staged-api.rs [hunk 6]
```diff
@@ -16,12 +14,11 @@ use staged_api::*;
...
 #[stable(feature = "rust1", since = "1.0.0")]
-#[cfg_attr(staged, rustc_const_stable(feature = "rust1", since = "1.0.0"))]
-// ^ should trigger error with or without the attribute
+#[cfg_attr(stable, rustc_const_stable(feature = "rust1", since = "1.0.0"))]
 impl const MyTrait for Stable {
-    fn func() { //~ ERROR trait methods cannot be stable const fn
-
-    }
+    //[stable]~^ ERROR trait implementations cannot be const stable yet
+    //[unstable]~^^ ERROR implementation has missing const stability attribute
```
Comment (jhpratt) [inline src/test/ui/rfc-2632-const-trait-impl/staged-api.rs:20]: This ensures that `impl const Trait` will require a const stability attribute once it's stabilized, just like everything else (#90998).

**Comment by camelid** [comment]
> Also included in this PR is a change of the default const stability level to **UN**stable. Previously, an item that was `const` but did not explicitly state it was unstable was implicitly stable.

Does this break anything that was already (potentially accidentally) stable? I thought it wasn't possible to have a `const` function in the stdlib without a const stability attribute.

**Comment by jhpratt** [comment]
There were a couple of methods that didn't have it, but they were intentional. That was fixed in #90998. This just changes the fallback to be unstable as a safety net, so to speak.

File: compiler/rustc_const_eval/src/transform/check_consts/mod.rs [hunk 7]
```diff
@@ -98,22 +96,39 @@ pub fn is_const_stable_const_fn(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
     // Const-stability is only relevant for `const fn`.
     assert!(tcx.is_const_fn_raw(def_id));
 
-    // Functions with `#[rustc_const_unstable]` are const-unstable.
+    // A function is only const-stable if it has `#[rustc_const_stable]` of it the trait it belongs
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/transform/check_consts/mod.rs:99]:
```suggestion
    // A function is only const-stable if it has `#[rustc_const_stable]` or if the trait it belongs
```

File: compiler/rustc_interface/src/passes.rs (hunk 3 above)
Comment (oli-obk) [reply compiler/rustc_interface/src/passes.rs:984]: If possible it should be done in `check_mod_unstable_api_usage`, if not, then call it next to the invocation further up.

File: compiler/rustc_passes/src/stability.rs (hunk 5 above)
Comment (oli-obk) [reply compiler/rustc_passes/src/stability.rs:926]: Should be increadibly cheap, but if you are using the output twice, this is better.

This function may go away anyway if it's possible to merge the logic into the general stability checking.

**Review by oli-obk — CHANGES_REQUESTED** [review]
Can we now remove `// build-fail` from some tests?

File: compiler/rustc_interface/src/passes.rs (hunk 3 above)
Comment (jhpratt) [reply compiler/rustc_interface/src/passes.rs:984]: Do you mean:

1. move the call to `check_const_impl_trait` to that function, or
2. move the code from `impl ItemLikeVisitor for CheckStableConstImplTrait` to `impl Visitor for Checker`?

I can see both sides of this. On one hand, the former would be easier to remove once `impl const Trait` is stabilized. But on the other hand, the latter would be more efficient, I would presume.

File: compiler/rustc_const_eval/src/transform/check_consts/mod.rs (hunk 7 above)
Comment (jhpratt) [reply compiler/rustc_const_eval/src/transform/check_consts/mod.rs:99]: Good catch.

**Review by jhpratt — COMMENTED** [review]
And yes, `// build-fail` can likely be removed from some tests.

**Comment by oli-obk** [comment]
I would prefer solution 2, but I realize it may be a larger diff or require duplication, so i'll trust your judgement on that.

I don't think we'll ever remove this stability check, after all, we will keep adding new traits and possibly make more of them const. Just like we have the const fn stability checks even though const fn is stable

**Comment by jhpratt** [comment]
True, I was thinking of the part that emits the error, which is only part of the change. I'll look into (2).

**Comment by jhpratt** [comment]
Do you see the need for both the [`stability.rs`](https://github.com/rust-lang/rust/blob/f838a425e3134d036a7d9632935111a569ac7446/src/test/ui/rfc-2632-const-trait-impl/stability.rs) and the [`staged-api.rs`](https://github.com/rust-lang/rust/blob/f838a425e3134d036a7d9632935111a569ac7446/src/test/ui/rfc-2632-const-trait-impl/staged-api.rs) tests? They will do pretty much the same thing after this change.

**Comment by oli-obk** [comment]
Yea they could be merged in case there is something only tested in one of them

**Comment by jhpratt** [comment]
Test failure is expected. I need to implement a check for a const trait impl being unstable with the feature enabled, as right now it appears that the compiler doesn't take this into account with regard to unstable-in-stable checking. Aside from that, I've merged the check into the existing stability pass.

**Comment by jhpratt** [comment]
Status update: I'm trying to figure out how to get `RUSTC_LOG` to work, which would let me debug the existing situation in the compiler. For some reason I can't get it to work at all, as the log level is capped. Once I get that working, it shouldn't be too difficult to implement this check 🤞🏽

**Comment by oli-obk** [comment]
> Status update: I'm trying to figure out how to get `RUSTC_LOG` to work, which would let me debug the existing situation in the compiler. For some reason I can't get it to work at all, as the log level is capped. Once I get that working, it shouldn't be too difficult to implement this check 🤞🏽

You need to set `debug-logging = true` in config.toml

**Comment by jhpratt** [comment]
Already did that to no avail, unfortunately.

**Comment by oli-obk** [comment]
Note: this will only work for the libstd built with the changed compiler. The initial libstd build will still be done with a downloaded stage0 compiler, maybe that's what you're seeing?

**Comment by jhpratt** [comment]
In that case, how can I ensure it's built myself? `x clean` by itself isn't sufficient: I've tried.

**Comment by oli-obk** [comment]
My recommendation would be to create a ui test that reproduces the issue, and then only useuse logging on that test.

If you do want to log libcore, you just need to wait until it gets built after the compiler is built once. Logging will work there. You will still get lots of messages about diablerie log levels while building stage 0 libstd and while building the compiler, but after the compiler is built, you will get logging output

**Comment by jhpratt** [comment]
Alright. So after rebasing and getting `RUSTC_LOG` working, it looks like the problem is that the `const fn` is being passed to `is_const_stable_const_fn`. We then try to get the parent of that, which naturally isn't the trait. What needs to be done is checking the function calls inside of the current parameter. I'm currently looking into the best way to do this. I presume there's something similar already in the codebase.

**Comment by JohnCSimon** [comment]
@jhpratt 
Ping from triage: Can you please post the status of this PR? 
Thank you.

**Comment by jhpratt** [comment]
I still plan on working on it! I've been busy with some other things and plan on picking it back up in the near future.

**Comment by jhpratt** [comment]
I asked on Zulip if there was anyone interested in taking this PR over (no responses yet). I have other priorities in the compiler, and I know there are still some design decisions to be made for `impl const Trait`, so this isn't a blocker of anything.

If anyone would like to take this over, please say so! If not, I'll likely close this pull request, at least for the time being. Anyone is free to use the code in this PR in a future PR (or more generally under MIT or Apache-2.0).

**Comment by jhpratt** [comment]
Closing for the reason stated above.

@rustbot label -S-waiting-on-author +S-inactive

**Pull request closed without merging**